### DISCUSSION TITLE
layout and view classes are incomplete when you use a default component view in admin

### DISCUSSION THREAD

**Discussion by shoulders** [joomla/joomla-cms#32620] [Q&A]
### Steps to reproduce the issue
In the admin area, go to any components default view by clicking on it's primary link:

![image](https://user-images.githubusercontent.com/319997/108200303-3063d400-7116-11eb-991a-a59630596929.png)

When you load the components default view the following classes are incomplete in the  `<body>` tag declaration, you get instead:

- **view-**
- **layout-**

and as you can see the names of the **view** and **layout** are missing part of the class declaration which means you cannot do **'page / view / layout'** specific CSS styling when it is loaded as a default view.

If you load `https://localhost/platforms/joomla3/administrator/index.php?option=com_contact&view=contacts` then the **view** class is correctly formed as **view-contacts**. However the **layout** is still incomplete probably because the layout is default.

### Expected result
```html
<body class=" jsn-master tmpl-isis admin com_contact view-contacts layout-default task- itemid-0" data-basepath="/platforms/joomla3">
```

### Actual result
```html
<body class=" jsn-master tmpl-isis admin com_contact view- layout- task- itemid-0" data-basepath="/platforms/joomla3">
```

### System information (as much as possible)

Setting | Value
-- | --
 
PHP Built On | Windows NT E6540 10.0 build 19042 (Windows 10) AMD64
Database Type | mysql
Database Version | 5.5.5-10.4.10-MariaDB
Database Collation | utf8mb4_unicode_ci
Database Connection Collation | utf8mb4_general_ci
PHP Version | 7.3.12
Web Server | Apache/2.4.41 (Win64) OpenSSL/1.1.1c PHP/7.3.12
WebServer to PHP Interface | apache2handler
Joomla! Version | Joomla! 3.9.24 Stable [ Amani ] 12-January-2021 15:00 GMT
Joomla! Platform Version | Joomla Platform 13.1.0 Stable [ Curiosity ] 24-Apr-2013 00:00 GMT
User Agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.150 Safari/537.36

### Additional comments
This is an issue in Joomla 3 but I have checked in joomla 4 and get the following notes. I also again used **com_contacts** and the example

- When you click on the primary link nothing happens so to load up the default view I used the following direct link instead
  `joomla4/administrator/index.php?option=com_contact`
- on the default page i got the following body tag i intiall got the following `<body>` tag but with **view-sysinfo** which is new over Joomla 3 and again **layout-** + there is now **task-**
  `<body class=" jsn-master tmpl-isis admin com_admin view-sysinfo layout- task- itemid-0" data-basepath="/platforms/joomla3">`
- after the intial load of `joomla4/administrator/index.php?option=com_contact` I now get a different `<body>` tag
  `<body class="admin com_contact view- layout-default task-display">`
  this now has the layout show correctly with default but the view is now missing part of the class name
- i load contacts up through the menu `Sidemenu --> Components -->Contacts --> Contacts` I get the following `<body>` tag
  `<body class="admin com_contact view-contacts layout-default task-display">`

# Conclusion

What this means is there is some inconsistencies with how these class names are created but has been partially fixed in Joomla 4.

This is not a show stopper but definately should be one of those things fixed in Joomla 4 if not in Joomla 3.

**Comment by brianteeman** [comment]
Please test and confirm that you have he same issue when using the joomla admin template and not the one from joomlashine

**Comment by richard67** [comment]
I can reproduce it on current staging with the Isis template:

![2021-02-17_01](https://user-images.githubusercontent.com/7413183/108218744-4208a400-7135-11eb-9f22-ee9b093229e4.png)

**Comment by shoulders** [comment]
I have just double checked the Joomlashine template issue. It was their system plugin adding `jsn-master` and I have reported this to them. Here is me checking my results with the ISIS template but without the JSN system plugin running.

![image](https://user-images.githubusercontent.com/319997/108230112-513d2100-7138-11eb-9c2f-acd96abd719d.png)

Alsojust notices that there is `task-` showing incomplete like in Joomla 4.

**Comment by joomdonation** [comment]
Easy to handle if it is really needed. With com_contact for example, just add two lines of code below before this line https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_contact/controller.php#L42 and it should work as expected:

```php
$this->input->set('view', $view);
$this->input->set('layout', $layout);
```

**Comment by shoulders** [comment]
this issue is with all components, not just `com_contacts` so it is an issue with the Joomla core and possibly how it handles default routes and that it does not add these variables into the request before Joomla builds the `<body>` tag . i.e.
- the default views can be set in the controller `protected $default_view = 'contacts';`
- if no layout is present in the request then `default` is assumed.
- if there is no task it is assumed that the task is `display`

**Comment by joomdonation** [comment]
We can handle  it in display method of BaseController class. Adding these lines of code below after this line https://github.com/joomla/joomla-cms/blob/staging/libraries/src/MVC/Controller/BaseController.php#L619

```php
$this->input->set('view', $viewName);
$this->input->set('layout', $viewLayout);
```
And that should work for all components. But I'm unsure if it is safe to do that.

**Comment by brianteeman** [comment]
What is the problem you are trying to solve

**Comment by shoulders** [comment]
thats an easy one.

- I am writing my own component which has a dashboard, and this dashboard is the default view.
- I want to style the background of the dashboard view (only) to grey so it makes the buttons stand out better but i do not want to make the backgrounds of all my other views grey.
- i am using 1 stylesheet in the admin for all of my component

this changes the background globaly in the admin for my component
```css
body.com_qwupdateserver {    
    background: #f9f9f9;
}
```
the next logical step is to use a page specific CSS
```css
body.view-dashboard {    
    background: #f9f9f9;
}
```

Solutions
- implement the issue related to in this thread. (prefered)
- i could add a specific CSS or ID into the dashboard html (seems wrong way when i should be able to use the auto generate css from Joomla)
- i could use double css identifiers: `.view-dashboard` and `view-` (messy)

On a personal note i want all of my CSS and code to be neat and tidy and I dont like broken code. 😄

**Comment by shoulders** [comment]
What a waste of my time. I report something that is clearly a bug and it just gets moved to discussions with no explanation, brilliant.

@rdeutz Why did you close this and move it to discussions? Is this the new MO of Joomla, move issues to discussions and not tell anyone why?

**Reply by brianteeman** [reply]
I agree the email message that its closed  is not very friendly to say the least
